src/Microsoft.ML.Dnn/TensorTypeExtensions.cs

```diff
@@ -3,49 +3,477 @@
+                            for (int i = 0; i < size; i++) destination[i] = ((IConvertible) src[i]).ToString(culture);
+                            return;
+                        }
+                        case TF_DataType.TF_UINT64:
+                        {
+                            var src = (ulong*) tensor.buffer;
+                            for (int i = 0; i < size; i++) destination[i] = ((IConvertible) src[i]).ToString(culture);
+                            return;
+                        }
+                        case TF_DataType.TF_DOUBLE:
+                        {
+                            var src = (double*) tensor.buffer;
+                            for (int i = 0; i < size; i++) destination[i] = ((IConvertible) src[i]).ToString(culture);
+                            return;
+                        }
+                        case TF_DataType.TF_FLOAT:
+                        {
+                            var src = (float*) tensor.buffer;
+                            for (int i = 0; i < size; i++) destination[i] = ((IConvertible) src[i]).ToString(culture);
+                            return;
+                        }
+#endif
+                        default:
+                            throw new NotSupportedException();
+                    }
+                }
+            }
+        }
+
+        public static void CopyTo<T>(this Tensor tensor, Span<T> destination) where T : unmanaged
+        {
             unsafe
             {
-                var len = checked((int)tensor.size);
-                var src = (T*)tensor.buffer;
-                var span = new Span<T>(src, len);
-                span.CopyTo(values);
+                var len = checked((int) tensor.size);
+                //perform regular CopyTo using Span.CopyTo.
+                if (typeof(T).as_dtype() == tensor.dtype && tensor.dtype != TF_DataType.TF_STRING) //T can't be a string but tensor can.
+                {
+                    var src = (T*) tensor.buffer;
+                    var srcspan = new Span<T>(src, len);
+                    srcspan.CopyTo(destination);
+
+                    return;
+                }
+
+                if (len > destination.Length)
+                    throw new ArgumentException("Destinion was too short to perform CopyTo.");
+
+                //Perform cast to type <T>.
+                fixed (T* dst_ = destination)
+                {
+                    var dst = dst_;
+                    switch (tensor.dtype)
+                    {
+#if _REGEN
+                        %foreach supported_numericals_TF_DataType,supported_numericals,supported_numericals_lowercase%
+                        case TF_DataType.#1:
+                        {
+                            var converter = Converts.FindConverter<#3, T>();
+                            var src = (#3*) tensor.buffer;
+                            Parallel.For(0, len, i => *(dst + i) = converter(unchecked(*(src + i))));
```

Comment:
Have you measured the perf of this? In the past, I've seen a tight loop inside a `Parallel.For` take longer than just a single loop. This is because `Parallel.For` needs to start new threads, and do coordination between the threads. When the work inside the loop is small, adding threading just makes this slower.

Reply:
Tensorflow.NET and NumSharp use `Parallel.For` extensively. Entirely up to you.

Reply:
I would want to see a perf comparison that proves `Parallel.For` is faster before I would use it in a tight loop.

Reply:
You are right, changing it to a for loop.
Benchmark: https://pastebin.com/raw/iALC8XAT

Reply:
Kidding? Linear always win?